src/data/zsqrtd/basic.lean

```diff
@@ -324,6 +331,36 @@ lemma norm_eq_one_iff {x : ℤ√d} : x.norm.nat_abs = 1 ↔ is_unit x :=
   exact this.1
 end⟩
 
+lemma is_unit_iff_norm_is_unit {d : ℤ} (z : ℤ√d) : is_unit z ↔ is_unit z.norm :=
+by rw [int.is_unit_iff_nat_abs_eq, norm_eq_one_iff]
+
+lemma norm_eq_one_iff' {d : ℤ} (hd : d ≤ 0) (z : ℤ√d) : z.norm = 1 ↔ is_unit z :=
+by rw [←norm_eq_one_iff, ←int.coe_nat_inj', int.nat_abs_of_nonneg (norm_nonneg hd z),
+  int.coe_nat_one]
+
+lemma norm_eq_zero_iff {d : ℤ} (hd : d < 0) (z : ℤ√d) : z.norm = 0 ↔ z = 0 :=
+begin
+  split,
+  { intro h,
+    rw [ext, zero_re, zero_im],
+    rw [norm_def, sub_eq_add_neg, mul_assoc] at h,
+    have left := mul_self_nonneg z.re,
+    have right := neg_nonneg.mpr (mul_nonpos_of_nonpos_of_nonneg hd.le (mul_self_nonneg z.im)),
+    obtain ⟨ha, hb⟩ := (add_eq_zero_iff_eq_zero_of_nonneg left right).mp h,
+    split; apply eq_zero_of_mul_self_eq_zero,
+    { exact ha },
+    { rw [neg_eq_zero, mul_eq_zero] at hb,
+      exact hb.resolve_left hd.ne } },
+  { rintro rfl, exact norm_zero }
+end
+
+lemma norm_eq_of_associated {d : ℤ} (hd : d ≤ 0) {x y : ℤ√d} (h : associated x y) :
```

Comment:
can't this be an iff?

Reply:
but norms are equal ±

Reply:
ahh, or is there not enough api? was just trying to prove it and realised there's nothing like multiplicative norm

Reply:
It's not true:
```lean
example : ∃ x y : ℤ√-3, x.norm = y.norm ∧ ¬associated x y :=
begin
  use [⟨1, 1⟩, ⟨1, -1⟩, by norm_num [zsqrtd.norm_def]],
  rintro ⟨u, hu⟩,
  rw [zsqrtd.ext] at hu,
  obtain ⟨h1, h2⟩ := hu,
  simp only [neg_mul_eq_neg_mul_symm, mul_one, one_mul, zsqrtd.mul_im, zsqrtd.mul_re] at h1 h2,
  rw ←eq_sub_iff_add_eq at h1,
  rw [h1, sub_neg_eq_add, add_comm, add_assoc, ←sub_eq_zero] at h2,
  have : 4 * (u : ℤ√-3).im = -2,
  { rw [←sub_eq_zero, ←h2],
    ring },
  apply_fun int.nat_abs at this,
  rw int.nat_abs_mul at this,
  have : int.nat_abs 4 ∣ int.nat_abs (-2),
  { rw ←this, apply dvd_mul_right },
  norm_num at this,
end 
```

Reply:
Right, I forgot that some ℤ√d aren't rings of integer of ℚ√d... sorry for the hassle!